### PR TITLE
backport from v3 to v2 MockNetworkURLProtocol and RequestRecorderURLProtocol files

### PR DESCRIPTION
**Pull request by codifilo**
# Submit a pull request

## CLA

- [X] I have signed the [Stream CLA](https://docs.google.com/forms/d/e/1FAIpQLScFKsKkAJI7mhCr7K9rEIOpqIDThrWxuvxnwUq2XkHyG154vQ/viewform) (required).
- [X] The code changes follow best practices
- [X] Code changes are tested (add some information if not applicable)

## Description of the pull request
The goal of this PR is to back-port the implementations of `MockNetworkURLProtocol` and `RequestRecorderURLProtocol` from v3 to v2.

The main difference between `MockNetworkURLProtocol` v2 and v3 is that in the `canInit(with:)` v2 method compares endpoint paths but v3 compares the whole URL so now  in order to use `MockNetworkURLProtocol` it’s necessary to know the whole URL for each endpoint. The v3 `ApiClient` implementation extracted this functionality to a `RequestEncoder` protocol that is accessible from the outside but in v2 this functionality is coupled and hidden in the `Client`. Therefore I exposed this functionality in `Client` in a method called `encodeRequest(for:)`

After I completed the back-port I noticed that a couple of tests would fail sometimes. The reason is that the new `MockNetworkURLProtocol` compares URL's but two URL's can be semantically identical but syntactically different.  Example: `https//a.b.c/d?e=f&g=h` and `https//a.b.c/d?g=h&e=f`. There is no guarantee that encoding a URL for the same endpoint twice would result in the same absolute URL string because it uses dictionaries to hold query items and  traverse order doesn't matter in dictionaries.

To address this in the `MockNetworkURLProtocol` I modified the key in the responses dictionary to compare significant parts of the URL by replacing the whole URL key for the URL scheme, host, port and path. Please note that this PR only addresses this issue in v2 but not in v3 so future tests in v3 can fail because of this. Please let me know if you want me to fix the issue also in the v3 class implementation.

### CODE REVIEW

**Comment by codifilo** [comment]
The two tests that were affected by the URL comparison issue I described above are failing now in the CI server. Right now this is the usual "it works on my machine" case as those tests pass quickly all the time on my macbook so I'm investigating why these tests fail in the CI server now. I'll post my findings here.

**Comment by codifilo** [comment]
I simplified even more the URL comparison to not take into account query items and now the tests pass in the CI server

**Comment by codifilo** [comment]
I see the Danger check was not successful: `For your GitHub repo, you need to expose: DANGER_GITHUB_API_TOKEN`
@VojtaStavik do I need to do some additional configuration in my forked repo?

**Comment by b-onc** [comment]
@codifilo That's expected from forks, no worries about danger

**Comment by codifilo** [comment]
I'm closing this PR as the commits are not sufficiently split. My approach was to commit as much as possible but not breaking the build or making tests to fail but the result it's hard to follow what's going on so it's hard to review the PR.
I will create a new one so the commits properly documents the changes in a way that it's easier to review.
